mllib/src/test/scala/org/apache/spark/ml/tree/impl/TreeSplitUtilsSuite.scala

```diff
@@ -0,0 +1,280 @@
+      // Choose split
+      val (split, stats) =
+        SplitUtils.chooseOrderedCategoricalSplit(statsAggregator, featureIndex, featureIndex)
+      // Verify that split has the expected left-side/right-side categories
+      val expectedRightCategories = Range(0, featureArity)
+        .filter(c => !expectedLeftCategories.contains(c)).map(_.toDouble).toArray
+      split match {
+        case s: CategoricalSplit =>
+          assert(s.featureIndex === featureIndex)
+          assert(s.leftCategories === expectedLeftCategories)
+          assert(s.rightCategories === expectedRightCategories)
+        case _ =>
+          throw new AssertionError(
+            s"Expected CategoricalSplit but got ${split.getClass.getSimpleName}")
+      }
+      validateImpurityStats(Entropy, labels, stats, expectedLeftStats, expectedRightStats)
+    }
+
+    // Test a single split: The left side of our split should contain the two points with label 0,
+    // the left side of our split should contain the five points with label 1
+    val values = Array(0, 0, 1, 2, 2, 2, 2)
+    val labels1 = Array(0, 0, 1, 1, 1, 1, 1).map(_.toDouble)
+    testHelper(values, labels1, expectedLeftCategories = Array(0.0),
+      expectedLeftStats = Array(2.0, 0.0), expectedRightStats = Array(0.0, 5.0))
+
+    // Test a single split: The left side of our split should contain the three points with label 0,
+    // the left side of our split should contain the four points with label 1
+    val labels2 = Array(0, 0, 0, 1, 1, 1, 1).map(_.toDouble)
+    testHelper(values, labels2, expectedLeftCategories = Array(0.0, 1.0),
+      expectedLeftStats = Array(3.0, 0.0), expectedRightStats = Array(0.0, 4.0))
+  }
+
+  test("chooseOrderedCategoricalSplit: return bad stats if we should not split") {
+    // Construct categorical data
+    val featureIndex = 0
+    val values = Array(0, 0, 1, 2, 2, 2, 2)
+    val featureArity = values.max + 1
+    val labels = Array(1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0)
+    // Construct DTStatsAggregator, compute sufficient stats
+    val metadata = TreeTests.getMetadata(numExamples = values.length, numFeatures = 1,
+      numClasses = 2, Map(featureIndex -> featureArity), unorderedFeatures = Some(Set.empty))
+    val statsAggregator = getAggregator(metadata, values, labels, featureSplits = Array.empty)
+    // Choose split, verify that it's invalid
+    val (_, stats) = SplitUtils.chooseOrderedCategoricalSplit(statsAggregator, featureIndex,
+        featureIndex)
+    assert(!stats.valid)
+  }
+
+  test("chooseUnorderedCategoricalSplit: basic case") {
+    val featureIndex = 0
+    // Construct data for unordered categorical feature
+    // label: 0 --> values: 1
+    // label: 1 --> values: 0, 2
+    // label: 2 --> values: 2
+    // Expected split: feature value 1 on the left, values (0, 2) on the right
+    val values = Array(1, 1, 0, 2, 2)
+    val featureArity = values.max + 1
```

Comment:
In order to make the test more strict, can you increase the `featureArity`, `numExamples` and `numClasses` ? e.g., featureArity = 6 and numExamples = 10 and numClasses = 5

Reply:
@WeichenXu123 thanks for the feedback! Definitely agree that the test is a little weak right now.

IMO it's mainly weak due to the low feature arity (there only three possible splits, so the right one could be picked by chance). I think increasing the number of classes/examples substantially might make the test harder to reason about, but not opposed to that either - let me know what you think.

What about something like:

```
    val values = Array(0, 1, 2, 3, 2, 2, 4)
    val labels = Array(0.0, 0.0, 1.0, 1.0, 2.0, 2.0, 2.0)
    // label: 0 --> values: 0, 1
    // label: 1 --> values: 2, 3
    // label: 2 --> values: 2, 2, 4
    // Expected split: feature values (0, 1) on the left, values (2, 3, 4) on the right
```

This way we still test multiclass classification & test the split-selection logic more rigorously.

Reply:
I think it is OK. thanks!